src/ns-support.c

```diff
@@ -45,6 +45,63 @@
 #include "mountinfo.h"
 #include "cleanup-funcs.h"
 
+/**
+ * Flag indicating that a sanity timeout has expired.
+ **/
+static volatile sig_atomic_t sanity_timeout_expired = 0;
+
+/**
+ * Signal handler for SIGALRM that sets sanity_timeout_expired flag to 1.
+ **/
+static void sc_SIGALRM_handler(int signum)
+{
+	sanity_timeout_expired = 1;
+}
+
+/**
+ * Enable a sanity-check timeout.
+ *
+ * The timeout is based on good-old alarm(2) and is intended to break a
+ * suspended system call, such as flock, after a few seconds. The built-in
+ * timeout is primed for three seconds. After that any sleeping system calls
+ * are interrupted and a flag is set.
+ *
+ * The call should be paired with sc_disable_sanity_check_timeout() that
+ * disabled the alarm and acts on the flag, aborting the process.
+ **/
+static void sc_enable_sanity_timeout()
+{
+	sanity_timeout_expired = 0;
+	struct sigaction act = {.sa_handler = sc_SIGALRM_handler };
+	// NOTE: we are using sigaction so that we can explicitly control signal
+	// flags and *not* pass the SA_RESTART flag. The intent is so that any
+	// system call we may be sleeping on to get interrupted.
+	if (sigaction(SIGALRM, &act, NULL) < 0) {
+		die("cannot install signal handler for SIGALRM");
+	}
```

Comment:
This is more standard (eg, http://www.gnu.org/software/libc/manual/html_node/Sigaction-Function-Example.html):
```
struct sigaction act;
act.sa_handler = sc_SIGALRM_handler;
if (sigemptyset(&act.sa_mask) < 0) {
    die(...);
}
act.sa_flags = 0;
if (sigaction(SIGINT, &act, NULL) < 0) {
    die(...);
}
```

If you use this, you can drop the comment since it is clear the flag won't be set.